kubeflow/trainer/rhai/transformers.py

```diff
@@ -0,0 +1,407 @@
+                "step_total": state.max_steps,
+                "percent": 0.0,
+                "epoch": state.epoch if hasattr(state, "epoch") else 0,
+            }},
+            "time": {{
+                "started_sec": self.start_time,
+                "started_at": self._format_timestamp(self.start_time),
+                "updated_sec": self.start_time,
+                "updated_at": self._format_timestamp(self.start_time),
+            }},
+        }})
+
+    def on_step_end(self, args, state, control, **kwargs) -> None:
+        """Called at the end of each training step."""
+        current_time = time.time()
+        current_step = state.global_step
+        total_steps = state.max_steps
+
+        progress_pct = (current_step / total_steps * 100) if total_steps > 0 else 0
+
+        elapsed_sec = current_time - self.start_time if self.start_time else 0
+        remaining_sec = None
+        if current_step > 0 and elapsed_sec > 0:
+            estimated_total_time = elapsed_sec / (current_step / total_steps)
+            remaining_sec = estimated_total_time - elapsed_sec
+
+        remaining_str = (
+            self._format_duration(remaining_sec) if remaining_sec else "calculating..."
+        )
+        status_msg = (
+            f"Training in progress: {{progress_pct:.1f}}% complete, "
+            f"{{remaining_str}} remaining"
+        )
+
+        ProgressMetricsHandler.update_metrics({{
+            "status_message": status_msg,
+            "progress": {{
+                "step_current": current_step,
+                "step_total": total_steps,
+                "percent": round(progress_pct, 2),
+                "epoch": state.epoch,
+            }},
+            "time": {{
+                "elapsed_sec": round(elapsed_sec, 2),
+                "elapsed": self._format_duration(elapsed_sec),
+                "remaining_sec": round(remaining_sec, 2) if remaining_sec else None,
+                "remaining": self._format_duration(remaining_sec) if remaining_sec else None,
+                "updated_sec": current_time,
+                "updated_at": self._format_timestamp(current_time),
+            }},
+        }})
+
+    def on_log(self, args, state, control, logs=None, **kwargs):
+        """Called when Trainer logs metrics."""
+        if logs:
+            metrics_update = {{}}
+
+            # Track standard metrics
+            if "loss" in logs:
+                metrics_update["loss"] = logs["loss"]
+            if "learning_rate" in logs:
+                metrics_update["learning_rate"] = logs["learning_rate"]
+            if "train_samples_per_second" in logs:
+                metrics_update["throughput_samples_sec"] = logs["train_samples_per_second"]
+
+            # Track custom metrics
+            if self.custom_metrics_map:
+                for log_key, metric_name in self.custom_metrics_map.items():
+                    if log_key in logs:
+                        metrics_update[metric_name] = logs[log_key]
```

Comment:
question: can you check transformers always provide us with float values? Looking at, e.g. [transformers mlflow integration](https://github.com/huggingface/transformers/blob/v4.57.1/src/transformers/integrations/integration_utils.py#L1497), the value could be a pytorch value.

Maybe we should do something similar
 
```suggestion
                        v = logs[log_key]
                        if isinstance(v, (int, float)):
                            metrics_update[metric_name] = v
                        elif isinstance(v, torch.Tensor) and v.numel() == 1:
                            metrics_update[metric_name] = v.item()
                        else:
                            logger.warning(...some warning message)
```

Reply:
Oh, great catch ! 🙌  👀